File: api/v1/endpoints/rl.py
# -*- coding: utf-8 -*-
"""RL 训练 API 端点"""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from api.deps import get_database_manager
from api.v1.schemas.rl import (
    DailyReplayResponse,
    EvaluateProgressResponse,
    EvaluateRequest,
    EvaluateResultResponse,
    EvaluateTaskResponse,
    ModelInfo,
    ModelListResponse,
    TaskStatusResponse,
    TrainRequest,
    TrainResponse,
    TrainingProgressResponse,
)
from src.services.rl_service import RLService
from src.storage import DatabaseManager

router = APIRouter()

# 全局 RLService 实例（懒加载）
_rl_service: Optional[RLService] = None


def get_rl_service() -> RLService:
    """获取 RLService 单例"""
    global _rl_service
    if _rl_service is None:
        from rl.config import RLConfig
        config = RLConfig.from_env()
        db = DatabaseManager.get_instance()
        _rl_service = RLService(config, db)
    return _rl_service


@router.get("/models", response_model=ModelListResponse)
async def list_models(service: RLService = Depends(get_rl_service)):
    """获取已训练的模型列表"""
    models = service.get_models()
    return ModelListResponse(
        models=[
            ModelInfo(
                model_id=m["model_id"],
                algorithm=m["algorithm"],
                created_at=m["created_at"],
                metrics=m.get("metrics"),
            )
            for m in models
        ]
    )


@router.post("/train", response_model=TrainResponse)
async def start_training(
    request: TrainRequest, service: RLService = Depends(get_rl_service)
):
    """启动异步训练任务（支持 resume_from 断点续训）"""
    params = {}
    if request.resume_from is not None:
        params["resume_from"] = request.resume_from
    if request.algorithm is not None:
        params["default_algorithm"] = request.algorithm
    if request.episodes is not None:
        params["training_episodes"] = request.episodes
    if request.batch_size is not None:
        params["batch_size"] = request.batch_size
    if request.learning_rate is not None:
        params["learning_rate"] = request.learning_rate

    task_id = service.start_training(params)
    return TrainResponse(task_id=task_id, status="pending", message="训练任务已创建")


@router.get("/train/{task_id}/status", response_model=TaskStatusResponse)
async def get_training_status(
    task_id: str, service: RLService = Depends(get_rl_service)
):
    """查询训练任务状态"""
    status = service.get_task_status(task_id)
    if status is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return TaskStatusResponse(**status)


@router.post("/train/{task_id}/stop")
async def stop_training(
    task_id: str, service: RLService = Depends(get_rl_service)
):
    """停止训练任务（保存断点，可通过 resume_from 续训）"""
    success = service.stop_training(task_id)
    if not success:
        raise HTTPException(status_code=404, detail="Task not found")
    return {"message": "停止指令已发送，正在保存断点..."}


@router.post("/train/{task_id}/pause")
async def pause_training(
    task_id: str, service: RLService = Depends(get_rl_service)
):
    """暂停训练任务（当前 episode 结束后生效）"""
    success = service.pause_training(task_id)
    if not success:
        raise HTTPException(status_code=400, detail="任务不存在或不在运行中")
    return {"message": "暂停指令已发送"}


@router.post("/train/{task_id}/resume")
async def resume_training(
    task_id: str, service: RLService = Depends(get_rl_service)
):
    """恢复已暂停的训练任务"""
    success = service.resume_training(task_id)
    if not success:
        raise HTTPException(status_code=400, detail="任务不存在或未处于暂停状态")
    return {"message": "训练已恢复"}


@router.get("/train/{task_id}/progress", response_model=TrainingProgressResponse)
async def get_training_progress(
    task_id: str, service: RLService = Depends(get_rl_service)
):
    """获取训练进度（供前端轮询）"""
    progress = service.get_training_progress(task_id)
    if progress is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return TrainingProgressResponse(**progress)


@router.post("/evaluate", response_model=EvaluateTaskResponse)
async def evaluate_model(
    request: EvaluateRequest, service: RLService = Depends(get_rl_service)
):
    """启动异步评估任务（前端轮询 /evaluate/{task_id}/progress 获取进度）"""
    try:
        task_id = service.start_evaluate(request.model_id, request.stock_codes, request.max_days)
    except KeyError as e:
        raise HTTPException(status_code=404, detail=str(e))
    return EvaluateTaskResponse(task_id=task_id, status="pending", message="评估任务已创建")


@router.get("/evaluate/{task_id}/progress", response_model=EvaluateProgressResponse)
async def get_evaluate_progress(
    task_id: str, service: RLService = Depends(get_rl_service)
):
    """获取评估任务进度（任务完成时附带完整评估结果）"""
    progress = service.get_evaluate_progress(task_id)
    if progress is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return EvaluateProgressResponse(**progress)


@router.post("/evaluate/{task_id}/pause")
async def pause_evaluate(task_id: str, service: RLService = Depends(get_rl_service)):
    """暂停评估任务（当前交易日回放完成后生效）"""
    success = service.pause_evaluate(task_id)
    if not success:
        raise HTTPException(status_code=400, detail="任务不存在或不在运行中")
    return {"message": "暂停指令已发送"}


@router.post("/evaluate/{task_id}/resume")
async def resume_evaluate(task_id: str, service: RLService = Depends(get_rl_service)):
    """恢复已暂停的评估任务"""
    success = service.resume_evaluate(task_id)
    if not success:
        raise HTTPException(status_code=400, detail="任务不存在或未处于暂停状态")
    return {"message": "评估已恢复"}


@router.post("/evaluate/{task_id}/stop")
async def stop_evaluate(task_id: str, service: RLService = Depends(get_rl_service)):
    """终止评估任务（在下一个交易日回放前中断）"""
    success = service.stop_evaluate(task_id)
    if not success:
        raise HTTPException(status_code=400, detail="任务不存在或已结束")
    return {"message": "终止指令已发送"}


@router.get("/evaluate/{model_id}/daily/{stock_code}/{date}", response_model=DailyReplayResponse)
async def get_daily_replay(
    model_id: str,
    stock_code: str,
    date: str,
    service: RLService = Depends(get_rl_service),
):
    """获取单日回放数据"""
    result = service.evaluate_daily(model_id, stock_code, date)
    if result is None:
        raise HTTPException(status_code=404, detail="Model not found")
    return DailyReplayResponse(**result)


@router.delete("/models/{model_id}")
async def delete_model(
    model_id: str, service: RLService = Depends(get_rl_service)
):
    """删除模型"""
    success = service.delete_model(model_id)
    if not success:
        raise HTTPException(status_code=404, detail="Model not found")
    return {"message": "模型已删除"}

File: apps/dsa-web/src/api/rl.ts
import apiClient from './index';
import { toCamelCase } from './utils';
import type {
  TrainRequest,
  TrainResponse,
  TaskStatusResponse,
  TrainingProgressResponse,
  ModelListResponse,
  EvaluateRequest,
  EvaluateTaskResponse,
  EvaluateProgressResponse,
  DailyReplayResponse,
} from '../types/rl';

/**
 * RL 训练 API 封装
 * 后端端点: /api/v1/rl/*
 */
export const rlApi = {
  /** 启动训练任务（支持 resumeFrom 断点续训） */
  startTraining: async (request: TrainRequest): Promise<TrainResponse> => {
    const payload: Record<string, unknown> = {};
    if (request.algorithm) payload.algorithm = request.algorithm;
    if (request.episodes != null) payload.episodes = request.episodes;
    if (request.batchSize != null) payload.batch_size = request.batchSize;
    if (request.learningRate != null) payload.learning_rate = request.learningRate;
    if (request.resumeFrom) payload.resume_from = request.resumeFrom;
    const response = await apiClient.post('/api/v1/rl/train', payload);
    return toCamelCase(response.data);
  },

  /** 查询任务状态 */
  getStatus: async (taskId: string): Promise<TaskStatusResponse> => {
    const response = await apiClient.get(`/api/v1/rl/train/${taskId}/status`);
    return toCamelCase(response.data);
  },

  /** 轮询训练进度（含完整指标历史） */
  getProgress: async (taskId: string): Promise<TrainingProgressResponse> => {
    const response = await apiClient.get(`/api/v1/rl/train/${taskId}/progress`);
    return toCamelCase(response.data);
  },

  /** 暂停训练（当前 episode 结束后生效） */
  pauseTraining: async (taskId: string): Promise<{ message: string }> => {
    const response = await apiClient.post(`/api/v1/rl/train/${taskId}/pause`);
    return response.data;
  },

  /** 恢复已暂停的训练 */
  resumeTraining: async (taskId: string): Promise<{ message: string }> => {
    const response = await apiClient.post(`/api/v1/rl/train/${taskId}/resume`);
    return response.data;
  },

  /** 停止训练（保存断点，可通过 resumeFrom 续训） */
  stopTraining: async (taskId: string): Promise<{ message: string }> => {
    const response = await apiClient.post(`/api/v1/rl/train/${taskId}/stop`);
    return response.data;
  },

  /** 获取已训练模型列表 */
  getModels: async (): Promise<ModelListResponse> => {
    const response = await apiClient.get('/api/v1/rl/models');
    return toCamelCase(response.data);
  },

  /** 启动异步评估任务（返回 taskId；maxDays 抽样评估天数，0/不传=全量默认抽样100） */
  startEvaluate: async (request: EvaluateRequest): Promise<EvaluateTaskResponse> => {
    const payload: Record<string, unknown> = { model_id: request.modelId };
    if (request.stockCodes) payload.stock_codes = request.stockCodes;
    if (request.maxDays != null) payload.max_days = request.maxDays;
    const response = await apiClient.post('/api/v1/rl/evaluate', payload);
    return toCamelCase(response.data);
  },

  /** 轮询评估任务进度（任务完成时附带完整评估结果） */
  getEvaluateProgress: async (taskId: string): Promise<EvaluateProgressResponse> => {
    const response = await apiClient.get(`/api/v1/rl/evaluate/${taskId}/progress`);
    return toCamelCase(response.data);
  },

  /** 暂停评估任务（当前交易日回放完成后生效） */
  pauseEvaluate: async (taskId: string): Promise<{ message: string }> => {
    const response = await apiClient.post(`/api/v1/rl/evaluate/${taskId}/pause`);
    return response.data;
  },

  /** 恢复已暂停的评估任务 */
  resumeEvaluate: async (taskId: string): Promise<{ message: string }> => {
    const response = await apiClient.post(`/api/v1/rl/evaluate/${taskId}/resume`);
    return response.data;
  },

  /** 终止评估任务（在下一个交易日回放前中断） */
  stopEvaluate: async (taskId: string): Promise<{ message: string }> => {
    const response = await apiClient.post(`/api/v1/rl/evaluate/${taskId}/stop`);
    return response.data;
  },

  /** 获取单日逐笔决策回放数据 */
  getDailyReplay: async (
    modelId: string,
    stockCode: string,
    date: string
  ): Promise<DailyReplayResponse> => {
    const response = await apiClient.get(
      `/api/v1/rl/evaluate/${modelId}/daily/${stockCode}/${date}`
    );
    return toCamelCase(response.data);
  },

  /** 删除模型 */
  deleteModel: async (modelId: string): Promise<{ message: string }> => {
    const response = await apiClient.delete(`/api/v1/rl/models/${modelId}`);
    return response.data;
  },
};

export default rlApi;
